iothub/device/src/ClientPropertyCollection.cs

```diff
@@ -180,35 +181,72 @@ public virtual bool TryGetValue<T>(string componentName, string propertyName, ou
                     {
-                        // If the value is null, go ahead and return it.
-                        if (dictionaryElement == null)
-                        {
-                            propertyValue = default;
-                            return true;
-                        }
-
-                        // If the object is of type T or can be cast to type T, go ahead and return it.
-                        if (dictionaryElement is T valueRef
-                            || NumericHelpers.TryCastNumericTo(dictionaryElement, out valueRef))
+                        if (nestedDictionary.TryGetValue(propertyName, out object dictionaryElement))
                         {
-                            propertyValue = valueRef;
-                            return true;
+                            // If the value associated with the key is null, then return successfully (true) with the default value of the type <T> passed in.
+                            if (dictionaryElement == null)
+                            {
+                                propertyValue = default;
+                                return true;
+                            }
+
+                            // If the object is of type T or can be cast to type T, go ahead and return it.
+                            if (dictionaryElement is T valueRef
+                                || NumericHelpers.TryCastNumericTo(dictionaryElement, out valueRef))
+                            {
+                                propertyValue = valueRef;
+                                return true;
+                            }
                         }
                     }
                 }
                 else
                 {
-                    // If it's not, we need to try to convert it using the serializer.
-                    Convention.PayloadSerializer.TryGetNestedObjectValue<T>(componentProperties, propertyName, out propertyValue);
-                    return true;
+                    // If the ClientPropertyCollection was constructed by the SDK (eg. when retrieving the client properties)
+                    // then the componentProperties are retrieved as the json object that is defined in the PayloadConvention.
+                    // The required property value then needs to be deserialized accordingly.
+                    try
+                    {
+                        // First verify that the retrieved dictionary contains the component identifier { "__t": "c" }.
+                        // If not, then the retrieved nested dictionary is actually a root-level property of type map.
+                        if (Convention
+                            .PayloadSerializer
+                            .TryGetNestedObjectValue(componentProperties, ConventionBasedConstants.ComponentIdentifierKey, out string componentIdentifierValue)
+                            && componentIdentifierValue.Equals(ConventionBasedConstants.ComponentIdentifierValue))
```

Comment:
Preferably use == for string comparisons. Switch order if you use .Equals to avoid nullref exception.